<launch>
    <node name="cam_capture_node" pkg="vapaa_cruiser" type="cam_capture_node.py" output="screen">
        <param name="width" value="320"/>
        <param name="height" value="240"/>
        <param name="rate" value="30"/>
        <param name="flipMode" value="0"/>
    </node>
    <node name="serial_command_node" pkg="vapaa_cruiser" type="serial_command_node.py" output="screen">
        <param name="port" value="/dev/ttyUSB0"/>
        <param name="baud" value="115200"/>
    </node>
    <node name="apriltag_node" pkg="vapaa_cruiser" type="apriltag_node.py" output="screen">
        <param name="tagSize" value="171"/>
    </node>
    <node name="fsm_node" pkg="vapaa_cruiser" type="fsm_node.py" output="screen">
    </node>
    <node name="image_store_node" pkg="vapaa_cruiser" type="image_store_node.py" output="screen">
        <param name="uploadHost" value="https://commutag.agawork.tw"/>
    </node>
    <node name="follow_tag_node" pkg="vapaa_cruiser" type="follow_tag_node.py" output="screen">
        <param name="tagID" value="1"/>
        <param name="keepDist" value="700"/>
        <param name="distTolerance" value="100"/>
        <param name="minAngle" value="-0.78539815"/>
        <param name="maxAngle" value="0.78539815"/>
        <param name="speedScale" value="0.0005"/>
        <param name="angleScale" value="1.5"/>
    </node>
    <node name="depth_process_node" pkg="vapaa_cruiser" type="depth_process_node.py" output="screen"> 
        <param name="rate" value="30"/>
    </node>
    
    <node name="web_video_server" pkg="web_video_server" type="web_video_server" output="screen"> 
        <param name="port" value="8080"/>
    </node>

    <node name="apriltag_repub" pkg="image_transport" type="republish" output="screen" args="compressed in:=/apriltag/detected raw out:=/apriltag_repub"></node>
    <node name="depth_process_repub" pkg="image_transport" type="republish" output="screen" args="compressed in:=/depth_process/image raw out:=/depth_process_repub"></node>

    <include file="$(find realsense2_camera)/launch/rs_camera.launch">
        <arg name="align_depth" value="true"/>
        <arg name="color_width" value="320"/>
        <arg name="color_height" value="240"/>
        <arg name="depth_width" value="640"/>
        <arg name="depth_height" value="480"/>
        <arg name="unite_imu_method" value="linear_interpolation"/>
        <arg name="enable_gyro" value="true" />
        <arg name="enable_accel" value="true" />
    </include>
    <include file="$(find rosbridge_server)/launch/rosbridge_websocket.launch" > 
        <arg name="port" value="9090"/>
    </include>

    <node name="imu_filter_node" pkg="imu_filter_madgwick" type="imu_filter_node" output="screen"> 
        <param name="use_mag" value="false"/>
        <param name="publish_tf" value="false"/>
        <param name="world_frame" value="enu"/>
        <remap from="/imu/data_raw" to="/camera/imu"/>
        <remap from="/imu/data" to="/rtabmap/imu"/>
    </node>
    
    <include file="$(find rtabmap_ros)/launch/rtabmap.launch">
        <arg name="rtabmap_args" value="--delete_db_on_start --Optimizer/GravitySigma 0.3"/>
        <arg name="depth_topic" value="/camera/aligned_depth_to_color/image_raw"/>
        <arg name="rgb_topic" value="/camera/color/image_raw"/>
        <arg name="camera_info_topic" value="/camera/color/camera_info"/>
        <arg name="approx_sync" value="false"/>
        <arg name="wait_imu_to_init" value="true"/>
        <arg name="imu_topic" value="/rtabmap/imu" />
        <arg name="localization" value="true" />
        <arg name="rtabmapviz" value="false" />
    </include>
</launch>
